pkg/agent/cniserver/pod_configuration_windows.go

```diff
@@ -0,0 +1,104 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cniserver
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/types/current"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/interfacestore"
+	"github.com/vmware-tanzu/antrea/pkg/agent/util"
+)
+
+func (pc *podConfigurator) connectInterfaceToOVSAsync(ifConfig *interfacestore.InterfaceConfig, containerAccess *containerAccessArbitrator) error {
+	if containerAccess == nil {
+		return fmt.Errorf("container lock cannot be null")
+	}
+	ovsPortName := ifConfig.InterfaceName
+	expectedEp, ok := pc.ifConfigurator.getEndpoint(ovsPortName)
+	if !ok {
+		return fmt.Errorf("failed to found HNSEndpoint: %s", ovsPortName)
+	}
+	hostIfAlias := fmt.Sprintf("%s (%s)", util.ContainerVNICPrefix, ovsPortName)
+	containerID := ifConfig.ContainerID
+	go func() {
+		klog.Infof("Waiting for interface %s to be created", hostIfAlias)
+		err := wait.PollImmediate(time.Second, 60*time.Second, func() (bool, error) {
+			curEp, ok := pc.ifConfigurator.getEndpoint(ovsPortName)
+			if !ok {
+				return true, fmt.Errorf("failed to found HNSEndpoint: %s", ovsPortName)
+			}
+			if curEp.Id != expectedEp.Id {
+				klog.Warningf("detected HNEEndpoint update: %s, exit current thread", ovsPortName)
+				return true, nil
+			}
+			if !util.HostInterfaceExists(hostIfAlias) {
+				klog.Infof("Waiting for interface: %s to be created", hostIfAlias)
+				return false, nil
+			}
+			if err := pc.connectInterfaceToOVSInternal(ovsPortName, ifConfig); err != nil {
+				return true, fmt.Errorf("failed to connect to ovs for container %s: %v", containerID, err)
+			}
+			return true, nil
+		})
+		if err != nil {
+			klog.Errorf("failed to create OVS port for container %s: %v", containerID, err)
+		}
+	}()
+	return nil
+}
+
+// connectInterfaceToOVS connects an existing interface to ovs br-int.
+func (pc *podConfigurator) connectInterfaceToOVS(
+	podName string,
+	podNameSpace string,
+	containerID string,
+	hostIface *current.Interface,
+	containerIface *current.Interface,
+	ips []*current.IPConfig,
+	containerAccess *containerAccessArbitrator,
+) (*interfacestore.InterfaceConfig, error) {
+	// Use the outer veth interface name as the OVS port name.
+	ovsPortName := hostIface.Name
+	containerConfig := buildContainerConfig(ovsPortName, containerID, podName, podNameSpace, containerIface, ips)
+	hostIfAlias := fmt.Sprintf("%s (%s)", util.ContainerVNICPrefix, ovsPortName)
+	if util.HostInterfaceExists(hostIfAlias) {
+		return containerConfig, pc.connectInterfaceToOVSInternal(ovsPortName, containerConfig)
+	} else {
+		return containerConfig, pc.connectInterfaceToOVSAsync(containerConfig, containerAccess)
```

Comment:
Can we add comments to explain this is for containerd?

Do you think we should differentiate Docker and containerd, and change the implementation only for containerd?

Reply:
Sure, add the comments.

The change is for ContainerD. But the logic is enough generic.  
Antrea will select the proper OVS creation function by checking if host interface exists. In this way we could differentiate Docker and ContainerD.

If ContainerD changes the order of VNIC creation in the future. Assume one day ContainerD also creates the VNIC during CNI call like Docker, the current implementation will till works.

Reply:
Ok. Probably still add enough comments to explain these implementation considerations clear, so we wont forget later.

Reply:
Sure, add more details about the implementaion.